proof-of-spacetime.md

```diff
@@ -0,0 +1,198 @@
+        if isFault {
+            // faulty sector, need to post a fault to the chain and try to recover from it
+            return Fatal("Detected late fault")
+        }
+
+        inclusionProofs[n] = inclusionProof
+		sectorsSorted[i] = sectors[sector]
+    }
+
+    // Generate the snark
+    snark_proof := GeneratePoStSnark(sectorSize, challenges, sectorsSorted, inclusionProofs)
+
+    proof := PoStProof {
+        snark: snark_proof
+    }
+
+    return proof, faults
+}
+```
+
+#### Verification
+
+`VerifyPoSt` is the functional counterpart to `GeneratePoSt`. It takes all of `GeneratePoSt`'s output, along with those of `GeneratePost`'s inputs required to identify the claimed proof. All inputs are required because verification requires sufficient context to determine not only that a proof is valid but also that the proof indeed corresponds to what it purports to prove.
+
+```go
+// Verify a PoSt.
+func VerifyPoSt(sectorSize BytesAmount, sectors SectorSet, seed Seed, proof PoStProof, faults FaultSet) bool {
+    challenges := DerivePoStChallenges(seed, faults, sectorSize, len(sectors))
+    sectorsSorted := []
+
+    // Match up commitments with challenges
+    for challenge in challenges {
+        sectorsSorted[i] = sectors[challenge.Sector]
+    }
+
+    // Verify snark
+    VerifyPoStSnark(sectorSize, challenges, sectorsSorted)
+}
+```
+
+
+#### Types
+
+```go
+// The final proof stored on chain.
+type PoStProof struct {
+    snark []byte
+}
+```
+
+```go
+// The random challenge seed, provided by the chain.
+Seed [32]byte
+```
+
+```go
+type Challenge struct {
+    Sector Uint
+    Leaf Uint
+}
+```
+
+#### Challenge Derivation
+
+```go
+// Derive the full set of challenges for PoSt.
+func DerivePoStChallenges(seed Seed, faults FaultSet, sectorSize: Uint, sectorCount: Uint) [POST_CHALLENGES_COUNT]Challenge {
+    challenges := []
+
+    for n in 0..POST_CHALLENGES_COUNT {
+        attempt := 0
+        while challenges[n] == nil {
+            challenge := DerivePoStChallenge(seed, n, faults, attempt, sectorSize, sectorCount)
+
+            // check if we landed in a faulty sector
+            if !faults.Contains(challenge.Sector) {
+                // Valid challenge
+                challenges[n] = challenge
+            }
+            // invalid challenge, regenerate
+            attempt += 1
+        }
+    }
+
+    return challenges
+}
+
+// Derive a single challenge for PoSt.
+func DerivePoStChallenge(seed Seed, n Uint, attempt Uint, sectorSize Uint, sectorCount: Uint) Challenge {
+    n_bytes := WriteUintToLittleEndian(n)
+    data := concat(seed, n_bytes, WriteUintToLittleEndian(attempt))
+    challenge_bytes := blake2b(data)
+
+    sector_challenge := ReadUintLittleEndian(challenge_bytes[0..8])
+    leaf_challenge := ReadUintLittleEndian(challenge_bytes[8..16])
+
+    Challenge {
+        Sector: sector_challenge % sectorCount,
+        Leaf: leaf_challenge % (sectorSize / NODE_SIZE),
```

Comment:
Question: Does the protocol define `NODE_SIZE` anywhere? I know it to be 32, but that's from looking at the software.

Reply:
yeah it is defined in the zigzag spec